src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs

```diff
@@ -113,21 +113,25 @@ public class Frame<TContext> : Frame
                                 await FireOnCompleted();
                             }
 
-                            _application.DisposeContext(context, _applicationException);
-                        }
+                            // If _requestAbort is set, the connection has already been closed.
+                            if (Volatile.Read(ref _requestAborted) == 0)
+                            {
+                                ResumeStreams();
 
-                        // If _requestAbort is set, the connection has already been closed.
-                        if (Volatile.Read(ref _requestAborted) == 0)
-                        {
-                            ResumeStreams();
+                                if (_keepAlive)
+                                {
+                                    // Finish reading the request body in case the app did not.
+                                    await messageBody.Consume();
+                                }
 
-                            if (_keepAlive)
-                            {
-                                // Finish reading the request body in case the app did not.
-                                await messageBody.Consume();
                             }
 
+                            // ProduceEnd() must be called before _application.DisposeContext(), to ensure
+                            // HttpContext.Response.StatusCode is correctly set when
+                            // IHttpContextFactory.Dispose(HttpContext) is called.
                             await ProduceEnd();
```

Comment:
`EngineTests.NoResponseSentWhenConnectionIsClosedByServerBeforeClientFinishesSendingRequest` seems to be failing because `ProduceEnd` was moved outside the if block.